### PR TITLE
Fix an issue in an edge case where sinks are disposed while broadcasting

### PR DESCRIPTION
**Pull request by gitusp**
Hello,

I found that disposing a sink, which is registered as a target of broadcasting, while broadcasting breaks succeeding broadcasting. This test proves the case.
The reason why it happens is that `broadcast` relies on `forEach`, but `forEach` does not care if the length of the array is changed while iterating. So it skips following sinks when `sinks` is shortened with `splice` while broadcasting.

Making a lock for broadcasting and deferring removing sinks after broadcasting solves the problem. But there might be better solution for the problem. Feel free to request changes.

I also installed testing environment to write the case. But I think it should fit main maintainer's preference. So please request changes if needed, too.

Thank you.

### CODE REVIEW

**Comment by gitusp** [comment]
Oh! Sorry I thought that I was working on my repository 💦
I'll finish my work soon. Please read this after finishing 🙇

**Comment by gitusp** [comment]
OK! I finished working on this pull request.

**Comment by briancavalier** [comment]
Hey @usp, thanks for catching this.  It definitely seems like a problem we need to address.  It's awesome that you added a simple test case, too.

A few quick thoughts for discussion before digging into the review:

> Making a lock for broadcasting and deferring removing sinks after broadcasting solves the problem. But there might be better solution for the problem. Feel free to request changes.

A potentially simpler solution may be just to duplicate the `sinks` array (e.g., `sinks.slice()`) at the start of `broadcast`, and iterate over the copy.  It's possible you considered that and it won't work for some reason that isn't occurring to me right now.  Any thoughts?

> I also installed testing environment to write the case.

I very much appreciate this!  A contributor [recently ported `most/hold` to TypeScript](https://github.com/mostjs/hold/pull/36), where we ended up with a test setup we were all happy with.  The setup uses 2 tsconfig files and a mocha.opts file.  It would be great to align the testing setup here with that one.

That PR also included jscpd and eslint, but imho, it'd be fine to skip those in this PR.  That way it adds only the minimum amount of new tooling required to prove the bug is fixed.

What do you think?

File: src/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { tap, take, runEffects } from "@most/core";
+import { newDefaultScheduler } from "@most/scheduler";
+import { describe, it } from "mocha";
+import assert from "power-assert";
```
Comment (briancavalier) [inline src/index.test.ts]: Let's stick with [`@briancavalier/assert`](https://www.npmjs.com/package/@briancavalier/assert), since the test case is so nice and simple (👏 )

File: src/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { tap, take, runEffects } from "@most/core";
```
Comment (briancavalier) [inline src/index.test.ts]: Let's match the code style in `src/index.ts`: single quotes, no semis.  Thanks!

**Comment by gitusp** [comment]
Thanks @briancavalier! I fixed the code style of test and changed assertion library.

> A potentially simpler solution may be just to duplicate the sinks array (e.g., sinks.slice()) at the start of broadcast, and iterate over the copy. It's possible you considered that and it won't work for some reason that isn't occurring to me right now. Any thoughts?

That seems simple and robust implementation. The only concern I have is whether duplicating an array for each broadcast have a negative effect on performance.
Having said that, I feel managing a lock is naive and would like to avoid to manage it if it's possible.
Another option I came up with is to `filter` the `sinks` on `dispose`, instead of `splice`, and assign the returned array to `sinks`. It will avoid duplicating `sinks` for each `broadcast`. (But ultimately it depends on how you use the library. If disposing and registering a sink is more often than broadcasting, the performance will be worse than duplicating `sinks` for each `broadcast`.)
What do you think?

> I very much appreciate this! A contributor recently ported most/hold to TypeScript, where we ended up with a test setup we were all happy with. The setup uses 2 tsconfig files and a mocha.opts file. It would be great to align the testing setup here with that one.

I agree. I think it should be aligned while this pull request should contain the minimal set of environmental setup. I'll fix it soon 👍

**Comment by gitusp** [comment]
I did a performance check for these versions:

- original implementation
- lock version(current implementation on this pull request)
- slice version(just added `slice()` before `forEach` to the original implementation)

The more sinks are registered, the less significant the difference is.

```
$ TS_NODE_COMPILER_OPTIONS='{"module":"commonjs"}' npx ts-node ./perf.ts
slice version - 1 sinks x 11,573,595 ops/sec ±0.32% (88 runs sampled)
lock version - 1 sinks x 14,163,358 ops/sec ±0.24% (91 runs sampled)
original version - 1 sinks x 14,256,623 ops/sec ±0.55% (92 runs sampled)
slice version - 2 sinks x 6,338,636 ops/sec ±0.19% (90 runs sampled)
lock version - 2 sinks x 7,214,024 ops/sec ±0.43% (89 runs sampled)
original version - 2 sinks x 7,222,101 ops/sec ±0.65% (93 runs sampled)
slice version - 4 sinks x 3,352,072 ops/sec ±0.63% (86 runs sampled)
lock version - 4 sinks x 3,665,262 ops/sec ±0.33% (92 runs sampled)
original version - 4 sinks x 3,670,743 ops/sec ±0.30% (91 runs sampled)
slice version - 8 sinks x 1,704,628 ops/sec ±0.58% (89 runs sampled)
lock version - 8 sinks x 1,819,190 ops/sec ±0.29% (91 runs sampled)
original version - 8 sinks x 1,812,998 ops/sec ±0.34% (91 runs sampled)
```

Performance checking code used is here: https://github.com/mostjs/adapter/commit/cd1196d25ff6028a3281af48e1dab6aefa34864c

**Comment by TylorS** [comment]
Hey @gitusp thanks so much for all your work on this so far. We might be able to get more performance out of the slice version if we do `.slice(0)` with the starting index. https://jsperf.com/cloning-arrays/3

**Comment by gitusp** [comment]
Thanks @TylorS! I updated the code to use `.slice(0)` and ran the benchmark again.
Here's the result:

```
$ uname -a
Darwin usps-mbp.local 18.2.0 Darwin Kernel Version 18.2.0: Mon Nov 12 20:24:46 PST 2018; root:xnu-4903.231.4~2/RELEASE_X86_64 x86_64

$ TS_NODE_COMPILER_OPTIONS='{"module":"commonjs"}' npx ts-node ./perf.ts
slice version - 1 sinks x 11,647,627 ops/sec ±0.37% (89 runs sampled)
lock version - 1 sinks x 14,460,026 ops/sec ±0.85% (88 runs sampled)
original version - 1 sinks x 14,647,864 ops/sec ±0.28% (92 runs sampled)
slice version - 2 sinks x 6,392,741 ops/sec ±0.31% (90 runs sampled)
lock version - 2 sinks x 7,344,381 ops/sec ±1.17% (88 runs sampled)
original version - 2 sinks x 7,378,000 ops/sec ±0.55% (91 runs sampled)
slice version - 4 sinks x 3,428,169 ops/sec ±0.29% (88 runs sampled)
lock version - 4 sinks x 3,741,570 ops/sec ±0.23% (87 runs sampled)
original version - 4 sinks x 3,721,257 ops/sec ±0.38% (88 runs sampled)
slice version - 8 sinks x 1,756,359 ops/sec ±0.59% (88 runs sampled)
lock version - 8 sinks x 1,804,324 ops/sec ±0.46% (92 runs sampled)
original version - 8 sinks x 1,841,366 ops/sec ±0.77% (87 runs sampled)
```

Unfortunately, I am not sure whether adding `0` to `slice` made it faster.

The code used here is https://github.com/gitusp/adapter/commit/c347e544b50ac87c67eb149117fb00c9ed805dbe.

**Comment by gitusp** [comment]
I completed to move test configurations from cli options for `mocha` to `mocha.opts` and `tsconfig.json` in `test` dir. Let me know if I'm missing something.
Thank you.

**Comment by gitusp** [comment]
Hi @briancavalier! This is just a reminder. Do you have an opinion on this topic? I'm waiting for your kind reply.
Thank you.

**Comment by briancavalier** [comment]
Thanks for the nudge @gitusp, and I'm sorry for not being very responsive lately.

tl;dr I strongly favor adding a strategically placed `.slice()` and calling it done :)

I'm in favor of the slice approach's simplicity for a few reasons:

1. The simpler code will be easier for the core team and the community to maintain and contribute to.
1. The lock version is impressive in that it nearly matches the original performance-wise.  However, overall, I'm not concerned about the slice performance numbers.  If there are issues in real world usage, we can address them once they are known.

There's a potential optimization to the slice version that should help the (most likely common) case of 1 sink: only slice if sinks.length > 1.  However, I don't think we should bother with that either.  If a concrete need to address the perf appears, we can do it at that time.

**Comment by gitusp** [comment]
Thanks @briancavalier for taking the time to review.

Although I have done that performance optimization, I agree with your opinion.
The potential optimization idea sounds smarter! I have never came up with that idea.

I have pushed simple `.slice()` version just now 🚀
